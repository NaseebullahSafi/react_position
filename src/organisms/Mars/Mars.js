import React from "react";
import Proptypes from 'prop-types'
import robot from "../../assets/icons/chevron-right.svg"
import './Mars.scss';

const includes = (arrOfArrs, arr) => arrOfArrs.some(a => arr.every((v, i) => v === [a.x, a.y][i]));

const Grid = ({ matrix, rovers }) => (
    <ul className="mars mars__grid">
        {
            matrix.map((el, i) => el.map(j => (
                <li className={`mars__grid-item ${includes(rovers, j.arr) ? 'mars__grid-item--active' : ''}`} key={j.coord}>
                    <div className="coordinates">{j.coord}</div>
                    {
                        rovers.map((rover, i) => rover.x === j.arr[0] && rover.y === j.arr[1] && (
                            <img
                                key={i}
                                src={robot}
                                className={`rover-face rover-face--${rover.face}`}
                            />
                        ))
                    }
                </li>
            )))
        }
    </ul>
)

const Mars = ({ dimension, rovers }) => {
    const [X, Y] = dimension
    const matrix = Array(Y).fill().map(() => Array(X).fill('-'));
    const newMatrix = []

    for (let i = 0; i < X; i++) {
        for (let j = Y - 1; j >= 0; j--) {
            matrix[i][j] = {
                coord: `( ${j} , ${i} )`,
                arr: [j, i]
            }
        }
    }

    for (let i = X - 1; i >= 0; i--) {
        newMatrix.push(matrix[i])
    }

    return (
        <section id="mars" className="mars">
            <Grid matrix={newMatrix} rovers={rovers} />
        </section>
    )
}

Mars.defaultProps = {
    dimension: [6, 6],
    rovers: [{
        x: 0,
        y: 0,
        face: 'N',
        path: ''
    }]
}

Mars.propTypes = {
    dimension: Proptypes.array,
    rovers: Proptypes.array
}

export default Mars;