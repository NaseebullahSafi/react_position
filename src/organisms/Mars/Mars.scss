.mars {
    background-color: white;
    border-radius: 5px;
    box-shadow: 0 0 12px -6px #d0d4d6;
    margin: 0 10px;
    height: 600px;
    overflow: auto;
    width: 100%;

    &__grid {
        background-color: white;
        display: flex;
        flex-wrap: wrap;
        list-style: none;
        margin: 0;
        padding: 0;
        width: 100%;

        &-item {
            align-items: center;
            background-color: white;
            display: flex;
            height: 150px;
            justify-content: center;
            padding: 10px;
            position: relative;
            transition: all .25s ease-in;
            width: calc(100% / 6);
            border-bottom: 1px solid #e7eaeb;
            &:nth-child(n):not(:nth-child(6n)) {
                border-right: 1px solid #e7eaeb;
            }

            .coordinates {
                align-items: center;
                border: 1px solid #efefef;
                border-radius: 51px;
                color: #616a71;
                display: flex;
                font-weight: bolder;
                height: 25px;
                justify-content: center;
                min-width: 40px;
                padding: 3px 17px;
                position: absolute;
                right: 18px;
                top: 13px;
                transition: all .25s ease-in;
            }

            &--active {
                background-color: #2bc2a3;
                border-radius: 5px;
                box-shadow: 0 0 8px -3px #4a5e69;
                z-index: 1;

                .coordinates {
                    background-color: white;
                    color: #979ea6;
                    transform: scale(1.1);
                    box-shadow: 1px 1px 3px 0px #898c8e;

                    & ~ .rover-face {
                        transition: all .25s cubic-bezier(.17,.67,.44,1.04);
                        
                        &--N {
                            transform: rotate(270deg);
                        }

                        &--E {
                            transform: rotate(0deg);
                        }

                        &--S {
                            transform: rotate(90deg);
                        }

                        &--W {
                            transform: rotate(180deg);
                        }

                    }
                }
            }
        }
    }
}