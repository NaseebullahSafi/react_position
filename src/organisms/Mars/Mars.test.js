import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Mars from './Mars';

afterEach(cleanup)

describe('Mars Component', () => {
  test('Matches the snapshot', () => {
    const dimension = [6,6]
    const rovers = [{
      x: 0,
      y: 0,
      face: 'N',
      path: ''
  }]
    const { asFragment } = render(<Mars dimension={dimension} rovers={rovers} />);
    expect(asFragment()).toMatchSnapshot()
  })
})