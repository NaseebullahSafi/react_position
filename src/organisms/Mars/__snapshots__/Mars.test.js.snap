// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mars Component Matches the snapshot 1`] = `
<DocumentFragment>
  <section
    class="mars"
    id="mars"
  >
    <ul
      class="mars mars__grid"
    >
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 0 , 5 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 1 , 5 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 2 , 5 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 3 , 5 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 4 , 5 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 5 , 5 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 0 , 4 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 1 , 4 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 2 , 4 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 3 , 4 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 4 , 4 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 5 , 4 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 0 , 3 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 1 , 3 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 2 , 3 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 3 , 3 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 4 , 3 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 5 , 3 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 0 , 2 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 1 , 2 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 2 , 2 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 3 , 2 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 4 , 2 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 5 , 2 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 0 , 1 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 1 , 1 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 2 , 1 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 3 , 1 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 4 , 1 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 5 , 1 )
        </div>
      </li>
      <li
        class="mars__grid-item mars__grid-item--active"
      >
        <div
          class="coordinates"
        >
          ( 0 , 0 )
        </div>
        <img
          class="rover-face rover-face--N"
          src="chevron-right.svg"
        />
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 1 , 0 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 2 , 0 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 3 , 0 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 4 , 0 )
        </div>
      </li>
      <li
        class="mars__grid-item "
      >
        <div
          class="coordinates"
        >
          ( 5 , 0 )
        </div>
      </li>
    </ul>
  </section>
</DocumentFragment>
`;
